akka-typed/src/main/scala/akka/typed/ScalaProcess.scala

```diff
@@ -0,0 +1,538 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com/>
+ */
+package akka.typed
+
+import scala.concurrent.duration._
+import akka.{ actor ⇒ a }
+
+object ScalaProcess {
+  import ScalaProcess._
+
+  class RetriesExceeded extends RuntimeException
+
+  import language.implicitConversions
+  /**
+   * This implicit expresses that operations that do not use their input channel can be used in any context.
+   */
+  private implicit def nothingIsSomething[T, U](op: Operation[Nothing, T]): Operation[U, T] = op.asInstanceOf[Operation[U, T]]
+
+  /*
+   * Convenient access to the core operations, will be even nicer with Dotty implicit function values
+   */
+  sealed trait OpDSL extends Any {
+    type Self
+  }
+
+  object OpDSL {
+    private val _unit: Operation[Nothing, Null] = opUnit(null)(null: OpDSL { type Self = Nothing })
+    private def unit[S, Out]: Operation[S, Out] = _unit.asInstanceOf[Operation[S, Out]]
+
+    def loopInf[S]: NextLoopInf[S] = nextLoopInf.asInstanceOf[NextLoopInf[S]]
+    trait NextLoopInf[S] {
+      def apply[U](body: OpDSL { type Self = S } ⇒ Operation[S, U]): Operation[S, Nothing] = {
+        lazy val l: Operation[S, Nothing] = unit[S, OpDSL { type Self = S }].flatMap(body).flatMap(_ ⇒ l)
+        l
+      }
+    }
+    private object nextLoopInf extends NextLoopInf[Nothing]
+
+    def loop[S]: NextLoop[S] = nextLoop.asInstanceOf[NextLoop[S]]
+    trait NextLoop[S] {
+      def apply[Out](n: Int)(body: OpDSL { type Self = S } ⇒ Operation[S, Out]): Operation[S, List[Out]] = {
+        require(n > 0, "number of iterations must be positive")
+        def step(n: Int, acc: List[Out]): Operation[S, List[Out]] =
+          unit[S, OpDSL { type Self = S }]
+            .flatMap(body)
+            .flatMap {
+              case result if n == 1 ⇒ Return((result :: acc).reverse)
+              case result           ⇒ step(n - 1, result :: acc)
+            }
+        step(n, Nil)
+      }
+    }
+    private object nextLoop extends NextLoop[Nothing]
+
+    def apply[T]: Next[T] = next.asInstanceOf[Next[T]]
+
+    trait Next[T] {
+      def apply[U](body: OpDSL { type Self = T } ⇒ Operation[T, U]): Operation[T, U] =
+        unit[T, OpDSL { type Self = T }].flatMap(body)
+    }
+    private object next extends Next[Nothing]
+
+    trait NextStep[T] {
+      def apply[U](mailboxCapacity: Int, body: OpDSL { type Self = T } ⇒ Operation[T, U])(implicit opDSL: OpDSL): Operation[opDSL.Self, U] =
+        Call(Process("nextStep", Duration.Inf, mailboxCapacity, body(null)))
+    }
+    private[typed] object nextStep extends NextStep[Nothing]
+  }
+
+  /*
+   * Terminology:
+   *
+   *  - a Process has a 1:1 relationship with an ActorRef
+   *  - an Operation is a step that a Process takes and that produces a value
+   *  - Processes are concurrent, but not distributed: all failures stop the entire Actor
```

Comment:
So I guess this is aimed to work well with the new supervision scheme you've been envisioning?
If we'd plug persistence, a store would be not per operation but per "successful application of process", so a fialure to apply a process would result in no persist happening? Trying to see how I'd design persistence using these hm...